dbms/src/Flash/Pipeline/Schedule/TaskQueues/ResourceControlQueue.cpp

```diff
@@ -0,0 +1,265 @@
+}
+
+template <typename NestedQueueType>
+void ResourceControlQueue<NestedQueueType>::submit(std::vector<TaskPtr> & tasks)
+{
+    std::lock_guard lock(mu);
+    for (auto & task : tasks)
+    {
+        submitWithoutLock(std::move(task));
+    }
+}
+
+template <typename NestedQueueType>
+void ResourceControlQueue<NestedQueueType>::submitWithoutLock(TaskPtr && task)
+{
+    if unlikely (is_finished)
+    {
+        FINALIZE_TASK(task);
+        return;
+    }
+    const auto & query_id = task->getQueryId();
+    if unlikely (cancel_query_id_cache.contains(query_id))
+    {
+        cancel_task_queue.push_back(std::move(task));
+        return;
+    }
+
+    // name can be empty, it means resource control is disabled.
+    const std::string & name = task->getResourceGroupName();
+
+    auto iter = resource_group_task_queues.find(name);
+    if (iter == resource_group_task_queues.end())
+    {
+        auto task_queue = std::make_shared<NestedQueueType>();
+
+        task_queue->submit(std::move(task));
+        resource_group_infos.push({LocalAdmissionController::global_instance->getPriority(name), task_queue, name});
+        resource_group_task_queues.insert({name, task_queue});
+    }
+    else
+    {
+        iter->second->submit(std::move(task));
+    }
+    cv.notify_one();
+}
+
+template <typename NestedQueueType>
+bool ResourceControlQueue<NestedQueueType>::take(TaskPtr & task)
+{
+    assert(task == nullptr);
+    std::unique_lock lock(mu);
+    while (true)
+    {
+        if (popTask(cancel_task_queue, task))
+            return true;
+
+        while (!resource_group_infos.empty())
+        {
+            if unlikely (is_finished)
+            {
+                // resource_group_infos and resource_group_task_queues will be cleaned in destructor,
+                // and all Tasks in nested task queue will be drained in destructor of TaskQueue.
+                return false;
+            }
+
+            ResourceGroupInfo group_info = resource_group_infos.top();
+            const std::string & name = std::get<InfoIndexResourceGroupName>(group_info);
+            const auto & priority = LocalAdmissionController::global_instance->getPriority(name);
+            std::shared_ptr<NestedQueueType> & task_queue = std::get<InfoIndexPipelineTaskQueue>(group_info);
+
+            LOG_TRACE(
+                logger,
+                "trying to schedule task of resource group {}, priority: {}, is_finished: {}, task_queue.empty(): {}",
+                name,
+                priority,
+                is_finished,
+                task_queue->empty());
+
+            // When highest priority of resource group is less than zero, means RU of all resource groups are exhausted.
+            // Should not take any task from nested task queue for this situation.
+            if (LocalAdmissionController::isRUExhausted(priority))
+                break;
+
+            if (task_queue->empty() || !task_queue->take(task))
```

Comment:
When will task_queue->take return false?
it seems impossible that this is the case
Add a runtime check?

Reply:
Yes. It make sense to add a runtime check